src/twisted/conch/ssh/transport.py

```diff
@@ -1454,7 +1454,7 @@ def _ssh_KEXDH_INIT(self, packet):
         self.g, self.p = _kex.getDHGeneratorAndPrime(self.kexAlg)
         self._startEphemeralDH()
         sharedSecret = self._finishEphemeralDH(clientDHpublicKey)
-        h = sha1()
+        h = _kex.getHashProcessor(self.kexAlg)()
```

Comment:
It looks like this will allow any of these algorithms (from `_kex.py`):

* curve25519-sha256
* curve25519-sha256@libssh.org
* diffie-hellman-group-exchange-sha256
* diffie-hellman-group-exchange-sha1
* diffie-hellman-group14-sha1
* ecdh-sha2-nistp256
* ecdh-sha2-nistp384
* ecdh-sha2-nistp521

Looking at the implementations, these are all `sha1`, `sha256`, `sha384` or `sha512`. Perhaps the change fragment should mention the newly supported algorithms?

Reply:
Many thanks for the prompt review :) Much appreciated.

This is called only for "non-group key exchange" aka `kex._IFixedGroupKexAlgorithm` 

The only fixed group that is still supported "upstream" by Twisted is "diffie-hellman-group14-sha1"

diffie-hellman-group1-sha1 was removed in #747 ... but it is not hard to monkey patch Twisted and add back support for it. I am doing it for my production code.

The main change in this PR is to make it easy to monkey patch  support for

```
diffie-hellman-group14-sha256
diffie-hellman-group16-sha512
diffie-hellman-group18-sha512
```

I have created ticket https://twistedmatrix.com/trac/ticket/10202 for that... but I don't know if other devs are interested into adding legacy KEX algorithms to Twisted.